### ISSUE TITLE
Negative number of elements on array dimensions

### ISSUE THREAD

**Issue by cristianbuse** [WaynePhillipsEA/twinbasic#176]
tB allows negative number of elements on array dimensions. The following code runs with in tB (not in VBA):
```VBA
Module Test
    Public Sub Main()
        Dim arr() As Variant
        
        ReDim arr(0 to -1)
        ReDim arr(0 to -10, 0 to -1)
        
        Stop
    End Sub
End Module
```

I like that I can declare a zero length array with ```arr(0 to -1)``` because it is not allowed in VBA. In VBA I could only get a zero-length array with something like ```arr = Array()``` and if needed for multiple dimensions then something like [this](https://stackoverflow.com/a/64320373/8488913) would work. But, do we want this capability of declaring an array with no elements or do we stick with the expected functionality of raising a runtime error?

Regarding multiple dimensions it seems that we only need one of the dimensions to be declared ```0 to -1``` and then it doesn't matter what we type in the other dimensions (ex. ```0 to -10```) because the array will get recognized as a zero-length array. This doesn't look right to me even if we were to accept the ```0 to -1``` notation.

**Comment by WaynePhillipsEA** [comment]
>But, do we want this capability of declaring an array with no elements or do we stick with the expected functionality of raising a runtime error?

Good question.  My gut feeling is no, as it seems a bit cryptic.  However, I'll leave this open for others to comment before fixing up to match VB.

**Comment by wqweto** [comment]
Btw this
```
Dim arr() As Variant
ReDim arr(0 to 0) As String
```
. . . is compile error in VBx but has not effect in TB (does not change element's type) which is harmless IMO.

**Comment by Kr00l** [comment]
> Btw this
> 
> ```
> Dim arr() As Variant
> ReDim arr(0 to 0) As String
> ```
> 
> . . . is compile error in VBx but has not effect in TB (does not change element's type) which is harmless IMO.

Harmless yes and no.
Imagine you write a module in tB which can also be used in VBx (for whatever reason) it would cause incompatibility as not interchange able.

So my stomage says to restrict to VBx rules.

**Comment by wqweto** [comment]
> Imagine you write a module in tB which can also be used in VBx (for whatever reason) it would cause incompatibility as not interchange able.

If you need source compatible implementation you would compile it in the more restrictive IDE not in the better/newer one.

For instance when I write stored procedures for MSSQL if I need these to be version compatible I test these on sql2000 not on sql2019 which has more extended features in the T-SQL but the language is backwards compatible with sql2000.

**Comment by Kr00l** [comment]
Maybe put this "option" whether to allow/disallow in the compiler optimization settings?
Which is then beneath the array overflow check setting.
The default would be then yes or no, depend on discussion.

But this would enable more flexibility.

**Comment by wqweto** [comment]
Yes, there has been some discussion about implementing "string mode" in TB but honestly that would be an enormous waste of development time IMO.

The compatibility has to be preserved in the other direction -- take existing VBx code, copy/paste it in TB and expect it to work without surprises like in VB.Net.

**Comment by cristianbuse** [comment]
> The compatibility has to be preserved in the other direction -- take existing VBx code, copy/paste it in TB and expect it to work without surprises (surprises like in VB.Net).

I agree. I would keep things as expected and raise a runtime error for a declaration of ```(x To y)``` where x is larger than y. It's rare that one would need a multidimensional array with no elements and for a one dimensional array we already have ```Array()```.

**Comment by bclothier** [comment]
FWIW, we do have the `Erase` statement to empty an array. That would be far more idiomatic than cryptic negative numbers. Statically creating an empty array should be a compile-time error.

**Comment by cristianbuse** [comment]
The ```Erase``` statement clears an array as if it was uninitialized. A zero-length array (as returned by an ```=Array()``` call is different because it has valid dimensions (```LBound```/```UBound``` do not raise an error) but it simply has no elements. Some people prefer using a zero-length array so that they don't need to check if the array is initialized before using a ```For Each``` loop for example. However, statically creating something like this is indeed too cryptic and not compatible.

In VBA, a line like ```Dim arr(0 To -5)``` does raise a compile-time error while in tB it doesn't. So this should be definitely fixed.

Of course, if we have something like ```Dim arr(): ReDim arr(x To y)``` there's nothing to do until runtime. But, how about ```Dim arr(): ReDim arr(0 To -5)``` which does not raise a compile-time error in VBA, only a runtime one. Would it make sense to raise a compile-time error in tB if the values are literals? It would seem like a nice addition.

**Comment by EduardoVB** [comment]
I vote to allow a way to initialize arrays with no elements. It is not too uncommon to need that.
In situations where you need to build a list of elements without knowing beforehand how many elements there will be, and if there will be any element at all, it is useful to start with a zero element array.
Using Array() or Split ("") functions, forces you to use Variant or String type arrays, so these workarounds are useful but limited.
Of course there are other ways to solve that situation, having a variable with the current index initialized to -1, but that cause to make the code more complex when it could be avoided.
It is one of the things that VB6 lacks IMO. 
But only Ubound = LBound -1 should be allowed, other negative values makes no sense.

**Comment by Kr00l** [comment]
So I agree with EduardoVB that arr(x To y) where y is -1 of x can be allowed to introduce a no-element array, for some reasons. But if  x - y is < -1 then an error should be thrown no matter what.
However, I still vote that this should be "allowed" explicitly via a compiler setting. So via such a setting it can be ensured that there are no arrays ReDim'ed by accident without any elements. (which is like in VBx, except for Array() etc. )

**Comment by WaynePhillipsEA** [comment]
I dislike the cryptic -1 syntax, and I don't think we'd be doing the BASIC language justice by allowing it.  Let's not do that, though I would be open to consider a more descriptive syntax.

**Comment by FullValueRider** [comment]
twinbasic has introduced assignment on initialisation/declaration and constructors for objects.  It would seem a reasonable request to extend such thinking to allow arrays (and collections and arraylists etc) to also be initialised at the point of declaration.  At which point the negative index issue goes away.

**Comment by WaynePhillipsEA** [comment]
Fixed in v0.10.871, thanks for reporting.

Please open a separate issue to discuss the potential zero-element array syntax.  I think the suggestion by @FullValueRider seems sensible.
